File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        ones = 0
        ans = 0
        
        l = 0
        
        for r in range(len(nums)):
            if nums[r] == 1:
                ones += 1
            while r-l+1 - ones > k:
                if nums[l] == 1:
                    ones -= 1
                l += 1
        
            ans = max(ans,r-l+1)
        
        return ans
            

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        //max length of subarray with atmost k zeros
        
        int ans = 0;
        
        int i = 0;
        int j = 0;
        
        while(i < nums.size() && j < nums.size()){
            if(nums[i] == 1){
                i++;
            }
            else if(nums[i] == 0 && k > 0){
                k--;
                i++;
            }
            else{
                if(nums[j] == 0){
                    k++;
                }
                j++;
            }
            ans = max(ans,i-j);
        }
        return ans;
    }
};